### PR TITLE
Fix issue #26: Cleanup and simplify code logic

### PR DESCRIPTION
**Pull request by jiere**
See #26 

Major change:
1. For `kubectl` and `kind`, using Github-hosted runners' provisioned version.
2. Simplify code logic, remove some unnecessary ENV options.
3. Update README.md and GHA workflow test.yml accordingly.

After this PR be merged, will cleanup related code in kepler-action repo accordingly.

Signed-off-by: Jie Ren <jie.ren@intel.com>

### CODE REVIEW

**Comment by jiere** [comment]
@SamYuan1990 , Please help review, thanks.

**Comment by SamYuan1990** [comment]
I am not sure. @marceloamaral , if we are considering with local dev env step up usage.... we may have to block this PR or at least a change request here.
@jiere , if further considering with self host runner, after this PR, the self host runner have to maintainer their own kubectl and kind version?

**Comment by jiere** [comment]
>  if further considering with self host runner, after this PR, the self host runner have to maintainer their own kubectl and kind version?

Agree, prerequisite well-known or popular packages should be maintained by machine owner, only those project specific tools need be taken care of by project's GHA code, IMO.

**Comment by rootfs** [comment]
cc @vprashar2929

**Comment by SamYuan1990** [comment]
I suppose we need a discussion as do we need handle kubectl and kind install script or not. I am now open for it.

File: common.sh
Comment (SamYuan1990) [inline common.sh:23]: en.... I don't like force port number. I hope to keep it flexible.

File: kind/kind.sh
Comment (SamYuan1990) [inline kind/kind.sh:72]: sed has different behaviors on mac and linux. :-)

Comment (SamYuan1990) [reply kind/kind.sh:72]: as I don't want to block mac user use this repo to setup their own dev cluster at local.
but... it seems the changes here is about do we need flexibility on port number?

**Comment by vprashar2929** [comment]
> I suppose we need a discussion as do we need handle kubectl and kind install script or not. I am now open for it.

I agree that we need a consensus on this. Do we want to use local-dev-cluster only for CI env setup or for local env setup as well? 
IMO it is best that we have a way to handle these dependency installations inside the script as we refer to local-dev-cluster in many of our other repos([kepler](https://github.com/sustainable-computing-io/kepler/blob/main/hack/cluster-up.sh),kepler-operator, etc) as part of the local dev environment setup plus it avoids any manual interventions to deal with dependencies whenever we need a quick setup to test kepler. 
I think we can modify the script to make sure that if it is referred to in GHA then it should not care about handling kubectl and kind installation. @SamYuan1990 @jiere Wdyt?

File: common.sh
Comment (jiere) [reply common.sh:23]: The reason why I chose to fix the port here is that if remain such flexibility, then we have a very ugly sed scripts to replace the port info inside the local-registry.yml.
As you could see in current PR diff, I also cleaned up there to make code cleaner:)
Flexibility often comes with complexity:)  So we may do trade-off. 
Some of the options are not so important to users, such as the registry port, for local-dev use and CI use, IMO, fix to 5001 is acceptable.

File: kind/kind.sh
Comment (jiere) [reply kind/kind.sh:72]: Exactly, why not we fix the content of both kind.yml and local-registry.yml :)

**Comment by jiere** [comment]
> Do we want to use local-dev-cluster only for CI env setup or for local env setup as well? 

My answer is both. But the point here is that prerequisite package setup should be one-shot job, it should be handled by runner admin or dev cluster owner. IMO, such kind/kubectl installation job is similar as Golang setup.

> IMO it is best that we have a way to handle these dependency installations inside the script as we refer to local-dev-cluster in many of our other repos [kepler](https://github.com/sustainable-computing-io/kepler/blob/main/hack/cluster-up.sh),kepler-operator, etc) as part of the local dev environment setup plus it avoids any manual interventions to deal with dependencies whenever we need a quick setup to test kepler. 

I agree that we should give convenience for developers to quickly setup local dev cluster and do quick try on Kepler. 
The reason I chose to make change here is that it is no so good to fix version of kind/kubectl, or even Golang here, since they are hot community and have quick version evolving, fix version means high risk on interoperation issues. 
IMO the best practice is to simply let developers to keep them up-to-date, unless our project has specific version limitation dependency on specific tools.

File: common.sh
Comment (SamYuan1990) [reply common.sh:23]: if we have to flexible with port number, any better solution than sed?

**Comment by SamYuan1990** [comment]
> > I suppose we need a discussion as do we need handle kubectl and kind install script or not. I am now open for it.
> 
> I agree that we need a consensus on this. Do we want to use local-dev-cluster only for CI env setup or for local env setup as well? IMO it is best that we have a way to handle these dependency installations inside the script as we refer to local-dev-cluster in many of our other repos([kepler](https://github.com/sustainable-computing-io/kepler/blob/main/hack/cluster-up.sh),kepler-operator, etc) as part of the local dev environment setup plus it avoids any manual interventions to deal with dependencies whenever we need a quick setup to test kepler. I think we can modify the script to make sure that if it is referred to in GHA then it should not care about handling kubectl and kind installation. @SamYuan1990 @jiere Wdyt?

> Do we want to use local-dev-cluster only for CI env setup or for local env setup as well?

Both.

> I think we can modify the script to make sure that if it is referred to in GHA then it should not care about handling kubectl and kind installation. 

@marceloamaral what do you think if we turn this repo from an empty linux to an environment same with today's GHA dependence?
as historical speaking, @marceloamaral is the most sponsor for local dev cluster at very beginning.

IMO, 
- It supports GHA, as current our pipeline we have.
- It supports Mac OS, as I am Mac user. 
- as currently it may able to start from a prune linux OS, which means a self hosted instance can free from kubectl and kind version. @jiere it's looks wired at my point of view, as you request flexible for multiple OS support,  but on the other hand... this PR is opened. ref to https://github.com/sustainable-computing-io/kepler-action/issues/50#issuecomment-1613085232, as we want to cover more CI tools, and if lucky, personally I hope to reuse currently code, hence I hope to keep robust, and avoid this repo force binding with GHA.

With a possible case, just assume this repo running on a self hosted instance, maybe triggered by GHA or not.
if we hard code as port number as 5001 which means for one time, we only have a single job can run. otherwise port conflicts. Hence further we need more flexibility and robust but not hard code.
May be further the cluster named kind is also a limitation, maybe the port and cluster name can be refined with a "random" value as `$job_number+kind` for port and `$job_number` as port.
And currently sed is a common way to replace the value. :-) if there any other good suggestion, please let me know.

**Comment by SamYuan1990** [comment]
As for today, I keep my standard to review this PR as the same for https://github.com/sustainable-computing-io/local-dev-cluster/pull/9/files if it doesn't break CI, then I can accept with merge it.
But for further consideration, I suppose this PR or parts of this PR reduced our robust and flexibility.  And in further if we considering with https://github.com/sustainable-computing-io/kepler-action/issues/50#issuecomment-1613085232 we have to add them back?(maybe not code back but feature back)

**Comment by jiere** [comment]
Thanks @SamYuan1990 's comments above, let me clarify two things:
1. This PR has no conflict with my earlier raised [issue](https://github.com/sustainable-computing-io/kepler-action/issues/50), in the mid-term or long-term, we should support more OS types of runners other than current Ubuntu. Current PR is discussing another topic, for any runner OS, shall we local-dev-cluster repo code take care of those prerequisite package installation stuffs? In my point of view, no matter which OS we are running on the runner, such package installation stuffs should be taken care of by the runner admin, but not our code. Take Github-hosted runner as example, Microsoft/Github guys are working hard to maintain their runners, even create [repo](https://github.com/actions/runner-images) for it, and also setup related [policies](https://github.com/actions/runner-images/tree/main#software-and-image-support). Today, we can use functions such as `action/setup-go`  in our GHA workflow yml just because they have written [code](https://github.com/actions/setup-go) for it, if no such code maintained by the runner admin, do we need to add the golong installation code in our local-dev-cluster repo also?

> if we hard code as port number as 5001 which means for one time, we only have a single job can run. otherwise port conflicts.

2. You persuaded me successfully for above reason, also thanks @vprashar2929 for your suggestions above, let me try if that change works and update the current diff later.

File: common.sh
Comment (jiere) [reply common.sh:23]: > Maybe we can use ex command like this:
> 
> ```
> ex -s -c "1,\$s/$kind_registry_name/${KIND_REGISTRY_NAME}/ge" -c "x" "${KIND_DIR}"/kind.yml
> ```
> 
> I think it is available on Mac and Linux


In my dev machine, the `ex` points to the same binary as `vim` :-)
$ ls -l /usr/bin/ex
lrwxrwxrwx 1 root root 20 Aug 24  2021 /usr/bin/ex -> /etc/alternatives/ex
$ ls -l /usr/bin/vim
lrwxrwxrwx 1 root root 21 Aug 24  2021 /usr/bin/vim -> /etc/alternatives/vim
$ ls -l /etc/alternatives/ex
lrwxrwxrwx 1 root root 18 Aug 24  2021 /etc/alternatives/ex -> /usr/bin/vim.basic
$ ls -l /etc/alternatives/vim
lrwxrwxrwx 1 root root 18 Aug 24  2021 /etc/alternatives/vim -> /usr/bin/vim.basic

Comment (vprashar2929) [reply common.sh:23]: Yes, it's the line editor for [vim](https://www.computerhope.com/unix/uex.htm#:~:text=On%20Unix%2Dlike%20operating%20systems,program%20written%20by%20Charles%20Haley.) 😅
It was new to me as well

**Comment by sthaha** [comment]
I have refactor PR that "cleans up" the setup script - https://github.com/sustainable-computing-io/local-dev-cluster/pull/31 
It would be nice to get your thoughts @SamYuan1990 @jiere

**Comment by vprashar2929** [comment]
> > sed is a common way to replace the value. :-) if there any other good suggestion, please let me know.
> 
> Just tried, `sed -i` worked well :-)

AFAIK `sed` has different implementations on Linux and macOS. On Linux, it seems to work fine but I have faced an issue on macOS. A related [PR](https://github.com/sustainable-computing-io/local-dev-cluster/pull/9) on this

File: common.sh
Comment (SamYuan1990) [reply common.sh:23]: em... well technologically maybe sed still one of best options?
I mean just when we replace string via shell.

**Comment by SamYuan1990** [comment]
> This PR has no conflict with my earlier raised https://github.com/sustainable-computing-io/kepler-action/issues/50, in the mid-term or long-term, we should support more OS types of runners other than current Ubuntu. Current PR is discussing another topic, for any runner OS, shall we local-dev-cluster [repo](https://github.com/actions/runner-images) code take care of those prerequisite package installation stuffs? In my point of view, no matter which OS we are running on the runner, such package installation stuffs should be taken care of by the runner admin, but not our code. Take Github-hosted runner as example, Microsoft/Github guys are working hard to maintain their runners, even create repo for it, and also setup related [policies](https://github.com/actions/runner-images/tree/main#software-and-image-support). Today, we can use functions such as action/setup-go in our GHA workflow yml just because they have written [code](https://github.com/actions/setup-go) for it, if no such code maintained by the runner admin, do we need to add the golong installation code in our local-dev-cluster repo also?

@jiere .... as mentioned in https://github.com/sustainable-computing-io/kepler-action/issues/50#issuecomment-1613085232 if we just considering with GHA, even if self host instance, then it works. but for other CI pipeines, as travis ... are those binaries are default settings? I am not sure...

**Comment by SamYuan1990** [comment]
@rootfs , what do you think of , do we need considering with other CI tools and keep kubectl and kind binary install today?

**Comment by marceloamaral** [comment]
I am open to utilizing pre-defined GHA (GitHub Actions) as long as we ensure consistency with the local cluster deployment. It is important for the developer to have a similar environment to test before pushing a commit.

Ultimately, we should use what makes the most sense for us.

**Comment by SamYuan1990** [comment]
> As for today, I keep my standard to review this PR as the same for https://github.com/sustainable-computing-io/local-dev-cluster/pull/9/files if it doesn't break CI, then I can accept with merge it. But for further consideration, I suppose this PR or parts of this PR reduced our robust and flexibility. And in further if we considering with [sustainable-computing-io/kepler-action#50 (comment)](https://github.com/sustainable-computing-io/kepler-action/issues/50#issuecomment-1613085232) we have to add them back?(maybe not code back but feature back)

hence recall myself, I prefer to merge this PR base on what we have today.
I will keep this PR open for review for this week.(@jiere , @rootfs , @vprashar2929 , @marceloamaral )
for port number we can fix it once we have a self hosted instance.

**Review by SamYuan1990 — APPROVED** [review]

File: kind/kind.sh
Comment (vprashar2929) [inline kind/kind.sh:46]: @jiere `sed -i` will not work on mac. The script fails with this change on mac

**Review by vprashar2929 — CHANGES_REQUESTED** [review]

File: common.sh
Comment (sthaha) [reply common.sh:23]: This issue is fixed in #31 by using `sed < manifest/yaml > generated/yaml`

File: kind/kind.sh
Comment (sthaha) [inline kind/kind.sh:67]: imho, we shouldn't remove this rather use the approach in #31  to validate if `kind exists` and the version is as expected.

File: kind/kind.sh
Comment (sthaha) [inline kind/kind.sh:84]: I think the function should adhere to single-responsiblity principle and only setup kind, not the registry. Use a different function (like it was earlier) to setup registry.

File: verify.sh
Comment (sthaha) [inline verify.sh:41]: Consider these 2 alternatives 

1. My preference - use `&& ||` 
```
kubectl cluster-info || { echo "failed .. "; return 1; }
```
in #31 I have a generic `die` method that exits 1

2. No need to use `if [` 

```
if ! kubectl cluster-info ; then 
  ... 
fi
```

File: kind/kind.sh
Comment (jiere) [reply kind/kind.sh:46]: Yes, I did see @sthaha 's suggestion above, let me change a new version and force-push diff again.

File: kind/kind.sh
Comment (jiere) [reply kind/kind.sh:67]: @sthaha , the key idea for this PR is not statically or ENV-based define the kind version, we just need to transfer the `kind` installation effort to runner admin, just as `Github-hosted runners` do. For developers, he/she needs to do one-shot installation of `kind` and `kubectl` when he setup his/her test machine at the very beginning.

File: kind/kind.sh
Comment (jiere) [reply kind/kind.sh:84]: Yeah, the reason why I added `_run_kind_registry` into `kind_up` is just to let the `kind` code in current common.sh align with  behavior of `microshift` code.
For current diff,  `_setup_kind` and `_run_kind_registry` are two steps for `_kind_up`, each has single-responsibility already; In common.sh, both `kind` and `microshift` cluster setup will have two similar steps: `_xxx_up` and `_wait_for_clusterReady`, not like current 3 steps for `kind` and 2 steps for `microshift` :-)

File: verify.sh
Comment (jiere) [reply verify.sh:41]: Cool, I like the first one :-) Let me change as you suggested then. Thanks.
The initial reason for me to change here is just removing the original console warning.

File: README.md
Comment (jiere) [reply README.md:21]: It seems that `export` grammar does not support such writings, let me change as:
`export CLUSTER_PROVIDER=kind or export CLUSTER_PROVIDER=microshift`, is this acceptable for you? @marceloamaral

File: kind/kind.sh
Comment (sthaha) [reply kind/kind.sh:67]: I agree about the installation part, but the validation for the versions must be performed in the script to ensure everyone stays on the same version. We can also add  `--no-validate  or --ci` flag to suppress unnecessary validations in CI.

File: kind/kind.sh
Comment (sthaha) [reply kind/kind.sh:91]: Speaking of which could this change wait for #31 which does a major overhaul of the scripts merge?

File: verify.sh
Comment (sthaha) [reply verify.sh]: Alternatively, since #31 is approved, it would be great if you could rebase the major changes on top of #31.

File: kind/kind.sh
Comment (sthaha) [inline kind/kind.sh:57]: can we move this validation to `verify.sh tools` instead? This allows everyone to validate if the versions of the binaries in path are good!

**Review by rootfs — APPROVED** [review]

**Comment by rootfs** [comment]
CI failed
```
./main.sh: line 74: CONFIG_OUT_DIR: unbound variable
Error: Process completed with exit code 1.
```

**Comment by jiere** [comment]
> CI failed
> 
> ```
> ./main.sh: line 74: CONFIG_OUT_DIR: unbound variable
> Error: Process completed with exit code 1.
> ```

Missed one line deletion... My bad, let me rebase again...

**Pull request merged**
